ort-custom-op/src/value.rs

```diff
@@ -0,0 +1,160 @@
+use crate::{
+    api::ElementType,
+    bindings::{OrtApi, OrtValue},
+};
+use anyhow::Result;
+use ndarray::{ArrayView, ArrayViewD};
+
+/// Enum over all currently supported input value types.
+#[derive(Debug)]
+pub enum Value<'a> {
+    TensorBool(ArrayViewD<'a, bool>),
+    TensorF32(ArrayViewD<'a, f32>),
+    TensorF64(ArrayViewD<'a, f64>),
+    TensorI16(ArrayViewD<'a, i16>),
+    TensorI32(ArrayViewD<'a, i32>),
+    TensorI64(ArrayViewD<'a, i64>),
+    TensorI8(ArrayViewD<'a, i8>),
+    TensorU16(ArrayViewD<'a, u16>),
+    TensorU32(ArrayViewD<'a, u32>),
+    TensorU64(ArrayViewD<'a, u64>),
+    TensorU8(ArrayViewD<'a, u8>),
+    TensorStr(ArrayViewD<'a, &'a str>),
+}
+
```

Comment:
Is there a reason this is an enum? Given sequences and maps are not on the horizon it should be fine to turn this into a struct.

Reply:
You are right that this `enum` is not strictly necessary, yet, but it makes life easier. If you look [here](https://github.com/cbourjau/ort-custom-op/blob/b6b12f33b708c5a32f1700a2d6f0c5b17ac7bee6/ort-custom-op/src/custom_op.rs#L188), we collect the input `Value`s into a `Vec`. These may be heterogeneous so we couldn't do that with the `ArrayViewD` (or a generic struct wrapping them) directly. However, we don't really need to collect in theory. We could implement a trait that assigns directly to a Tuple of the correct size for the operator. As a matter of fact, this is how it was done previously and it could still be done. However, with the new indirection around the ownership of the buffer that trait would become very complicated. It is easier to just collect these `Values` and then create the input tuple from that.

Indeed, I was thinking about `Sequence`s and `Map`s as well when picking this design. I didn't really explore it yet, but I'm not sure they need to be that far off and it would be a pitty to do another major refactor if we need them. Furthermore, there is also the other use case of an operator with heterogeneous variadic input. It is not enabled yet, but it would also work with this design. The compute function would directly take a `Value` as input in that case.